consensus/istanbul/backend/engine.go

```
@@ -395,39 +394,71 @@ func (sb *Backend) Prepare(chain consensus.ChainReader, header *types.Header) er
-		if additionalParentSeals != nil && additionalParentSeals.Size() != 0 {
-			logger.Trace("Combining additional seals with the parent aggregated seal", "additionalSeals", additionalParentSeals.String(), "num", number, "parentAggregatedSeal", parentAggregatedSeal.String())
-			// if we had any seals gossiped to us, proceed to add them to the
-			// already aggregated signature
-			if unionAggregatedSeal, err := istanbulCore.UnionOfSeals(parentExtra.AggregatedSeal, additionalParentSeals); err != nil {
-				logger.Error("Failed to create union of parent aggregated seal", "err", err)
-			} else {
-				// need to pass the previous block from the parent to get the parent's validators
-				// (otherwise we'd be getting the validators for the current block)
-				parentValidators := sb.getValidators(parent.Number.Uint64()-1, parent.ParentHash)
-				// only update to use the union if we indeed provided a valid aggregate signature for this block
-				if err := sb.verifyAggregatedSeal(parent.Hash(), parentValidators, unionAggregatedSeal); err != nil {
-					logger.Error("Failed to combine additional seals with parent aggregated seal.", "err", err)
+		logger = logger.New("parentAggregatedSeal", parentAggregatedSeal.String())
+		logger.Trace("Got ParentAggregatedSeal")
+		if seq != nil && seq.Cmp(header.Number) == 0 {
+			parentCommits := sb.core.ParentCommits()
+			logger = logger.New("cur_seq", sb.core.Sequence())
+			if parentCommits != nil && parentCommits.Size() != 0 {
+				logger = logger.New("numParentCommits", parentCommits.Size())
+				logger.Trace("Found commit messages from previous sequence to combine with ParentAggregatedSeal")
+				// if we had any seals gossiped to us, proceed to add them to the
+				// already aggregated signature
+				if unionAggregatedSeal, err := istanbulCore.UnionOfSeals(parentExtra.AggregatedSeal, parentCommits); err != nil {
+					logger.Error("Failed to combine commit messages with ParentAggregatedSeal", "err", err)
```

same here, why not return an error?

---

Same answer, we can get by with just using the signatures from the block in our chain